presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.util.ImmutableCollectors;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class TypeSignatureProvider
+{
+    private final boolean hasDependency;
+    private final Function<List<Type>, TypeSignature> typeSignatureSupplier;
+
+    public TypeSignatureProvider(TypeSignature typeSignature)
+    {
+        this.hasDependency = false;
+        this.typeSignatureSupplier = ignored -> typeSignature;
+    }
+
+    public TypeSignatureProvider(Function<List<Type>, TypeSignature> typeSignatureSupplier)
+    {
+        this.hasDependency = true;
+        this.typeSignatureSupplier = requireNonNull(typeSignatureSupplier, "typeSignatureSupplier is null");
+    }
+
+    public boolean hasDependency()
```

Comment:
Can you help me understand why you think this would be better called `isStructural`? Can you also share your concern about why you don't like the original name?

Generally, structural types means array, map, row, etc. I don't really think of function type as a structural type. Regardless, given structural type also includes array, map, row, `isStructural` doesn't seem like a good name?